### PR TITLE
ADC STM32: Various fixes and enhancements

### LINKED ISSUE #79909: stm32h7: adc: sequence of more than 8 channels does not work
**Describe the bug**
Starting an adc scan sequence of more than 8 channels does not work on STM32H7

Tested on nucleo_h723zg with adc_sequence sample

**To Reproduce**

1. create an overlay for the stm32h723 in the board folder of the adc_sequence sample (see below)
2. update the prj.conf to add logs and reduce extra samples with 
```
CONFIG_LOG=y
CONFIG_ADC_LOG_LEVEL_DBG=y
CONFIG_SEQUENCE_SAMPLES=1
```
4. go to samples/drivers/adc/adc_sequence and build the project with `west build -b nucleo_h723zg . -p`
5. flash to nucleo board with `west flash`

Example of overlay I used:
```

/ {
	zephyr,user {
	};

	aliases {
		adc0 = &adc1;
	};
};

&gpiod {
    status = "okay";
};

&gpioc {
    status = "okay";
};

&gpiob {
    status = "okay";
};

&clk_lsi {
	status = "okay";
};

&clk_hsi48 {
	status = "okay";
};

&clk_hsi {
	status = "okay";
	hsi-div = <1>;
};

&clk_hse {
	clock-frequency = <DT_FREQ_M(25)>;
	status = "disabled";
};

&pll {
	div-m = <4>;
	mul-n = <34>;
	div-p = <1>;
	div-q = <68>;	/* DIVQ at 68 to get 8MHz FDCAN clock */
	div-r = <2>;
	clocks = <&clk_hsi>;
	status = "okay";
};

&pll2 {
	status = "okay";
	clocks = <&clk_hsi>;
	div-m = <4>;
	mul-n = <12>;
	div-p = <4>;
	div-q = <2>;
	div-r = <2>;
};

&rcc {
	clock-frequency = <DT_FREQ_M(544)>;
	clocks = <&pll>;
	d1cpre = <1>;
	hpre = <2>;    /* HCLK: 272   MHz */
	d1ppre = <2>;  /* APB1: 136 MHz */
	d2ppre1 = <2>; /* APB2: 136 MHz */
	d2ppre2 = <2>; /* APB3: 136 MHz */
	d3ppre = <2>;  /* APB4: 136 MHz */
};


&adc1 {
	#address-cells = <1>;
	#size-cells = <0>;
	status = "okay";

	pinctrl-0 = <&adc1_inp11_pc1 &adc1_inp10_pc0 &adc1_inp17_pa1 &adc1_inp16_pa0
		&adc1_inp15_pa3 &adc1_inp14_pa2 &adc1_inp19_pa5 &adc1_inp18_pa4 
		&adc1_inp7_pa7 &adc1_inp3_pa6 &adc1_inp8_pc5 &adc1_inp4_pc4
		&adc1_inp5_pb1 &adc1_inp9_pb0 &adc1_inp6_pf12 &adc1_inp2_pf11>;
	pinctrl-names = "default";
	st,adc-clock-source = <ASYNC>;	/* PLL2_P @48MHz */
	st,adc-prescaler = <1>;
	st,adc-sequencer = <FULLY_CONFIGURABLE>;
	vref-mv = <2000>;

	channel@2 {
		reg = <2>;
		zephyr,gain = "ADC_GAIN_1";
		zephyr,reference = "ADC_REF_INTERNAL";
		zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
		zephyr,vref-mv = <2000>;
		zephyr,resolution = <16>;
	};

	channel@3 {
		reg = <3>;
		zephyr,gain = "ADC_GAIN_1";
		zephyr,reference = "ADC_REF_INTERNAL";
		zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
		zephyr,vref-mv = <2000>;
		zephyr,resolution = <16>;
	};

	channel@4 {
		reg = <4>;
		zephyr,gain = "ADC_GAIN_1";
		zephyr,reference = "ADC_REF_INTERNAL";
		zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
		zephyr,vref-mv = <2000>;
		zephyr,resolution = <16>;
	};

	channel@5 {
		reg = <5>;
		zephyr,gain = "ADC_GAIN_1";
		zephyr,reference = "ADC_REF_INTERNAL";
		zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
		zephyr,vref-mv = <2000>;
		zephyr,resolution = <16>;
	};

	channel@6 {
		reg = <6>;
		zephyr,gain = "ADC_GAIN_1";
		zephyr,reference = "ADC_REF_INTERNAL";
		zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
		zephyr,vref-mv = <2000>;
		zephyr,resolution = <16>;
	};

	channel@7 {
		reg = <7>;
		zephyr,gain = "ADC_GAIN_1";
		zephyr,reference = "ADC_REF_INTERNAL";
		zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
		zephyr,vref-mv = <2000>;
		zephyr,resolution = <16>;
	};

	channel@8 {
		reg = <8>;
		zephyr,gain = "ADC_GAIN_1";
		zephyr,reference = "ADC_REF_INTERNAL";
		zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
		zephyr,vref-mv = <2000>;
		zephyr,resolution = <16>;
	};

	channel@9 {
		reg = <9>;
		zephyr,gain = "ADC_GAIN_1";
		zephyr,reference = "ADC_REF_INTERNAL";
		zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
		zephyr,vref-mv = <2000>;
		zephyr,resolution = <16>;
	};

	channel@a {
	 	reg = <10>;
	 	zephyr,gain = "ADC_GAIN_1";
	 	zephyr,reference = "ADC_REF_INTERNAL";
	 	zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
	 	zephyr,vref-mv = <2000>;
	 	zephyr,resolution = <16>;
	 };
};
```

**Expected behavior**
Sequence should start and complete in a loop and analog values should be displayed

**Impact**
Sequence need to be split to be at most 8 channels long

**Logs and console output**
Here is the log I have with 9 channels:

```
[00:00:00.000,000] <dbg> adc_stm32: adc_stm32_channel_setup: Channel setup succeeded!
[00:00:00.000,000] <dbg> adc_stm32: adc_stm32_channel_setup: Channel setup succeeded!
[00:00:00.000,000] <dbg> adc_stm32: adc_stm32_channel_setup: Channel setup succeeded!
[00:00:00.000,000] <dbg> adc_stm32: adc_stm32_channel_setup: Channel setup succeeded!
[00:00:00.000,000] <dbg> adc_stm32: adc_stm32_channel_setup: Channel setup succeeded!
[00:00:00.000,000] <dbg> adc_stm32: adc_stm32_channel_setup: Channel setup succeeded!
[00:00:00.000,000] <dbg> adc_stm32: adc_stm32_channel_setup: Channel setup succeeded!
[00:00:00.000,000] <dbg> adc_stm32: adc_stm32_channel_setup: Channel setup succeeded!
[00:00:00.000,000] <dbg> adc_stm32: adc_stm32_channel_setup: Channel setup succeeded!
[00:00:01.002,000] <dbg> adc_stm32: adc_stm32_start_conversion: Starting conversion
[00:00:01.002,000] <dbg> adc_stm32: adc_stm32_isr: adc@40022000 ISR triggered.
```

When I only set 8 channels (commenting out one channel), I get this logs: in a loop:
```
[00:00:00.000,000] <dbg> adc_stm32: adc_stm32_channel_setup: Channel setup succeeded!
[00:00:00.000,000] <dbg> adc_stm32: adc_stm32_channel_setup: Channel setup succeeded!
[00:00:00.000,000] <dbg> adc_stm32: adc_stm32_channel_setup: Channel setup succeeded!
[00:00:00.000,000] <dbg> adc_stm32: adc_stm32_channel_setup: Channel setup succeeded!
[00:00:00.000,000] <dbg> adc_stm32: adc_stm32_channel_setup: Channel setup succeeded!
[00:00:00.000,000] <dbg> adc_stm32: adc_stm32_channel_setup: Channel setup succeeded!
[00:00:00.000,000] <dbg> adc_stm32: adc_stm32_channel_setup: Channel setup succeeded!
[00:00:00.000,000] <dbg> adc_stm32: adc_stm32_channel_setup: Channel setup succeeded!
- adc@40022000, channel 2, 1 sequence samples:
- - 1015 = 0mV
- adc@40022000, channel 3, 1 sequence samples:
- - 751 = 0mV
- adc@40022000, channel 4, 1 sequence samples:
- - 8 = 0mV
- adc@40022000, channel 5, 1 sequence samples:
- - 480 = 0mV
- adc@40022000, channel 6, 1 sequence samples:
- - 1002 = 0mV
- adc@40022000, channel 7, 1 sequence samples:
- - 9 = 0mV
- adc@40022000, channel 8, 1 sequence samples:
- - 13 = 0mV
- adc@40022000, channel 9, 1 sequence samples:
- - 10 = 0mV
ADC sequence reading [1]:
[00:00:01.002,000] <dbg> adc_stm32: adc_stm32_start_conversion: Starting conversion
[00:00:01.002,000] <dbg> adc_stm32: adc_stm32_isr: adc@40022000 ISR triggered.
[00:00:01.002,000] <dbg> adc_stm32: adc_stm32_isr: adc@40022000 ISR triggered.
[00:00:01.002,000] <dbg> adc_stm32: adc_stm32_isr: adc@40022000 ISR triggered.
[00:00:01.002,000] <dbg> adc_stm32: adc_stm32_isr: adc@40022000 ISR triggered.
[00:00:01.002,000] <dbg> adc_stm32: adc_stm32_isr: adc@40022000 ISR triggered.
[00:00:01.002,000] <dbg> adc_stm32: adc_stm32_isr: adc@40022000 ISR triggered.
[00:00:01.002,000] <dbg> adc_stm32: adc_stm32_isr: adc@40022000 ISR triggered.
[00:00:01.002,000] <dbg> adc_stm32: adc_stm32_isr: adc@40022000 ISR triggered.
- adc@40022000, channel 2, 1 sequence samples:
- - 656 = 0mV
- adc@40022000, channel 3, 1 sequence samples:
- - 436 = 0mV
- adc@40022000, channel 4, 1 sequence samples:
- - 11 = 0mV
- adc@40022000, channel 5, 1 sequence samples:
- - 350 = 0mV
- adc@40022000, channel 6, 1 sequence samples:
- - 534 = 0mV
- adc@40022000, channel 7, 1 sequence samples:
- - 14 = 0mV
- adc@40022000, channel 8, 1 sequence samples:
- - 9 = 0mV
- adc@40022000, channel 9, 1 sequence samples:
- - 11 = 0mV
```

**Environment (please complete the following information):**

 - OS: Linux (Ubuntu 24.04)
 - Toolchain: Zephyr sdk 0.16.5
 - Release 3.7.0

**Additional context**

### PR DESCRIPTION
**Pull request by gautierg-st**
This PR contains various fixes and enhancements for the STM32 ADC driver:
- Add a log message when an overrun error occurs (with indications how to prevent the error)
- Simplify the oversampling function by adding a new property in device tree
- Simplify the DMA enable function and add support for DMA for STM32H7A/H7B and F1 series

Fixes #79909

### CODE REVIEW

File: drivers/adc/adc_stm32.c [hunk 1]
```diff
@@ -230,38 +230,16 @@ struct adc_stm32_cfg {
...
-	} else {
-		LL_ADC_REG_SetDataTransferMode(adc, LL_ADC_REG_DMA_TRANSFER_LIMITED);
-	}
-#elif defined(STM32H74X_ADC)
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32f1_adc)
+	LL_ADC_REG_SetDMATransfer(adc, LL_ADC_REG_DMA_TRANSFER_UNLIMITED);
+#elif defined(CONFIG_SOC_SERIES_STM32H7X) || defined(CONFIG_SOC_SERIES_STM32U5X)
+	/* H72x ADC3 and U5 ADC4 are different from the rest, but this call works also for them,
+	 * so no need to call their specific function
+	 */
```
Comment (erwango) [inline drivers/adc/adc_stm32.c:238]: Open: Does commenting LL shenanigans adds clarity to the reader here ?

**Review by erwango — COMMENTED** [review]
A nice improvement. Few open comments, but otherwise LGTM

File: drivers/adc/adc_stm32.c [hunk 2]
```diff
@@ -58,20 +58,26 @@ LOG_MODULE_REGISTER(adc_stm32);
...
-#define STM32F3X_ADC_V1_1
-#endif
-#endif
+#define STM32F3XX_ADC /* Other F3xx */
+#endif /* ADC5_V1_1 */
+#elif defined(CONFIG_SOC_SERIES_STM32H7X)
+#if defined(ADC_VER_V5_V90)
+#define STM32H72X_ADC /* H72x and H73x */
+#elif defined(ADC_VER_V5_X)
+#define STM32H74X_ADC /* H74x and H75x */
```
[the remark this answers is not in the document]
Comment (gautierg-st) [reply drivers/adc/adc_stm32.c:73]: I don't see how. ADC_VER_V5_X is common to all U5, and there's nothing that would group U5 and H74x without the other H7.

File: drivers/adc/adc_stm32.c (hunk 1 above)
Comment (gautierg-st) [reply drivers/adc/adc_stm32.c:238]: For a reader who would be aware that these ADCs are different (and I include myself in it), it mentions the fact that this is done on purpose, that the peculiarities of these ADC were taken into account.

File: dts/bindings/adc/st,stm32-adc.yaml [hunk 3]
```diff
@@ -104,5 +104,18 @@ properties:
       - <NOT_FULLY_CONFIGURABLE>: Not fully configurable sequencer
       - <FULLY_CONFIGURABLE>: Fully configurable sequencer
 
+  st,adc-oversampler:
+    type: int
+    required: true
+    enum:
+      - 0 # OVERSAMPLER_NONE
+      - 1 # OVERSAMPLER_MINIMAL
+      - 2 # OVERSAMPLER_EXTENDED
```
[the remark this answers is not in the document]
Comment (gautierg-st) [reply dts/bindings/adc/st,stm32-adc.yaml:113]: Will do. And will do the same for the sequencer while I'm at it

[the remark this answers is not in the document]
Comment (gautierg-st) [reply dts/bindings/adc/st,stm32-adc.yaml]: Done

File: drivers/adc/adc_stm32.c [hunk 4]
```diff
@@ -793,10 +793,13 @@ static void dma_callback(const struct device *dev, void *user_data,
 
 	if (channel == data->dma.channel) {
 #if !DT_HAS_COMPAT_STATUS_OKAY(st_stm32f1_adc)
-		if (LL_ADC_IsActiveFlag_OVR(adc) || (status >= 0)) {
-#else
-		if (status >= 0) {
+		if (LL_ADC_IsActiveFlag_OVR(adc)) {
+			LL_ADC_ClearFlag_OVR(adc);
+			LOG_ERR("ADC overrun error occurred. Reduce clock source frequency, "
+				"increase prescaler value, or increase sampling times.");
```
[the remark this answers is not in the document]
Comment (gautierg-st) [reply drivers/adc/adc_stm32.c]: Done

**Review by erwango — DISMISSED** [review]

File: drivers/adc/adc_stm32.c [hunk 5]
```diff
@@ -734,8 +669,11 @@ static void adc_stm32_oversampling_ratioshift(ADC_TypeDef *adc, uint32_t ratio,
  * ratio is directly the sequence->oversampling (a 2^n value)
  * shift is the corresponding LL_ADC_OVS_SHIFT_RIGHT_x constant
  */
-static int adc_stm32_oversampling(ADC_TypeDef *adc, uint8_t ratio)
+static int adc_stm32_oversampling(const struct device *dev, uint8_t ratio)
 {
+	const struct adc_stm32_cfg *config = dev->config;
+	ADC_TypeDef *adc = (ADC_TypeDef *)config->base;
```
Comment (bjarki-andreasen) [inline drivers/adc/adc_stm32.c:675]: The typecast here is not needed, its already `ADC_TypeDef *`

File: include/zephyr/dt-bindings/adc/stm32_adc.h [hunk 6]
```diff
@@ -84,4 +84,15 @@
...
+ * @name STM32 ADC oversampler type
+ * This value is to set <st,adc-oversampler>
+ * Refer to the RefMan to know which type corresponds to which ADC.
+ * @{
+ */
+#define OVERSAMPLER_NONE	0	/* No oversampler */
+#define OVERSAMPLER_MINIMAL	1	/* 8 oversampling values available */
+#define OVERSAMPLER_EXTENDED	2	/* 1024 oversampling values available */
+/** @} */
+
```
Comment (bjarki-andreasen) [inline include/zephyr/dt-bindings/adc/stm32_adc.h:97]: Is this used? it seems this PR decided on using strings for `OVERSAMPLER_` and `_CONFIGURABLE` enums

Comment (bjarki-andreasen) [reply include/zephyr/dt-bindings/adc/stm32_adc.h:97]: looking a bit further, the defines are used in the driver, not in the devicetree, so I would move these inside the driver as they seemingly don't need to be exposed to the devicetree :)

Comment (gautierg-st) [reply include/zephyr/dt-bindings/adc/stm32_adc.h]: You're right, they're not needed here anymore. Moved into the driver

**Comment by gautierg-st** [comment]
Also applied string enum to st,adc-clock-source property that I missed earlier

**Review by erwango — DISMISSED** [review]

**Review by bjarki-andreasen — APPROVED** [review]
Nice work!

**Review by erwango — APPROVED** [review]

**Comment by benediktibk** [comment]
Could it be that there is an entry in the release notes/migration guide missing? Because changing the type of the properties to string is somewhat a breaking change. I just stumbled upon this. No big deal, but might be noteworthy in the documentation.

**Comment by kartben** [comment]
> Could it be that there is an entry in the release notes/migration guide missing? Because changing the type of the properties to string is somewhat a breaking change. I just stumbled upon this. No big deal, but might be noteworthy in the documentation.

+1. @gautierg-st please follow up on adding this to the migration guide.

**Comment by gautierg-st** [comment]
> > Could it be that there is an entry in the release notes/migration guide missing? Because changing the type of the properties to string is somewhat a breaking change. I just stumbled upon this. No big deal, but might be noteworthy in the documentation.
> 
> +1. @gautierg-st please follow up on adding this to the migration guide.

Done here #85643

**Pull request merged**